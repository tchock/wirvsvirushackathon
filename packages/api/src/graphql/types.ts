export type UserInfo = {
  iss: String;
  sub: String;
  aud: String[];
  iat: Number;
  exp: Number;
  azp: String;
  scope: String;
  permissions: String[];
};

export type ResolverContext = {
  user: UserInfo;
};
