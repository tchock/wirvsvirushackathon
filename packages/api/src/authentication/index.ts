import { AuthenticationError } from 'apollo-server-lambda';
import authenticate from './authenticate';

class NoAuthHeaderError extends AuthenticationError {}
class WrongTokenTypeError extends AuthenticationError {}
class InvalidTokenTypeError extends AuthenticationError {}

async function authenticateHeader(header) {
  if (!header) {
    throw new NoAuthHeaderError('No authorization header found');
  }

  const [tokenType, token] = header.split(' ');

  if (tokenType !== 'Bearer') {
    throw new WrongTokenTypeError('Wrong token type');
  }

  try {
    return { user: await authenticate(token) };
  } catch (e) {
    throw new InvalidTokenTypeError(e.message);
  }
}

function fakeAuth() {
  return {
    user: {
      iss: 'https://outofthebox.eu.auth0.com/',
      sub: 'github|38959977',
      aud: ['api.outofthebox.com', 'https://outofthebox.eu.auth0.com/userinfo'],
      iat: 1579344887,
      exp: 1579352087,
      azp: 'wiBC7LZCn4zmRVq5zEIMVuC9xkA9vRYJ',
      scope: 'openid profile',
      permissions: ['access'],
    },
  };
}

async function lambdaMiddleware({ event: { headers } }) {
  return fakeAuth();
  // return authenticateHeader(headers.Authorization || headers.authorization);
}

async function httpMiddleware(req) {
  return authenticateHeader(req.headers.Authorization || req.headers.authorization);
}

export { lambdaMiddleware, httpMiddleware };
