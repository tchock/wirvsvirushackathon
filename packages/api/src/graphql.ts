import { makeExecutableSchema } from "apollo-server-lambda"

const typeDef = /* GraphQL */ `
type Query {
  orders(audience: Audiences!): OrderList
  order(
    nodeId: String!
  ): Order
}

type Mutation {
  orderPickUp(pickUpCode: String!): Order # order status -> PICKED_UP
  orderDecline(nodeId: String!): Order # order status -> REJECTED
  orderAccept(nodeId: String!): Order # order status -> ACCEPTED
  orderPlace(order: OrderInput): Order # creates order
}

enum Audiences {
  CUSTOMER
  STORE
}

enum OrderStatus {
  ACCEPTED
  REJECTED
  PICKED_UP
}

interface Node {
  nodeId: String! # base64 PK::SK
}

type OrderList {
  # edges: [OrderEdge]!
  nodes: [Order]!
}

# type OrderEdge {
#   cursor: String!
#   node: Order!
# }

type Order implements Node {
  nodeId: String!
  pickUpCode: String!
  confirmedPickUpTime: String # ISO8601
  requestedPickUpTime: String # ISO8601
  store: Store!
  customer: Customer!
  orderStatus: OrderStatus!
  bundles: BundleList!
  shareLink: String!
}

type OrderInput {
  bundles: [BundleInput!]!
  store: String! # Store nodeId
  requestedPickUpTime: String # ISO8601
}

type BundleList {
  # edges: [BundleEdge]!
  nodes: [Bundle]!
}

# type BundleEdge {
#   cursor: String!
#   node: Bundle!
# }

type Bundle implements Node {
  nodeId: String! # NOT any PK + SK
  items: ItemList!
}

type BundleInput {
  items: ItemList!
}

type ItemList {
  # edges: [ItemEdge]!
  nodes: [Item]!
}

# type ItemEdge {
#   cursor: String!
#   node: Item!
# }

type Item {
  nodeId: String! # NOT any PK + SK
  quantity: Float!
  price: Float!
  name: String!
  unit: String!
}

interface User {
  type: Audiences!
}

type Store implements Node & User {
  nodeId: String!
  type: STORE
}

type Customer implements Node & User {
  nodeId: String!
  type: CUSTOMER
}
`

const resolvers = {

  // Query: {
  //   orders:  async (root, args, context) => {}
  // },

  // Order: {
  //   nodeId: nodeIdResovler
  // }

}

export const schema = makeExecutableSchema({ typeDefs: [typeDef], resolvers })
