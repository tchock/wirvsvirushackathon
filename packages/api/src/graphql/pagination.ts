export function asPaginationResolver(resolver) {
  return async (...args) => {
    const result = await resolver(...args);
    return {
      nodes: result,
      // TODO edges
      // TODO pageInfo
    };
  };
}
