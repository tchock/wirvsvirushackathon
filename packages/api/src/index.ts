import server from './apolloServer';

export const handler = server.createHandler({
  cors: {
    origin: '*',
  },
});
