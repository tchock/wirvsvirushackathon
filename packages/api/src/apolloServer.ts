import { ApolloServer } from 'apollo-server-lambda';
import { schema } from './graphql';
// import { lambdaMiddleware } from './authentication';
// import requestIdContext from './misc/requestIdContext';
// import logger from './misc/logger';

const server = new ApolloServer({
  schema,
  context: async (...args) => {
    // const { transactionId } = requestIdContext();
    const { headers } = args[0].event;
    const referer = headers.Referer || headers.referer;
    const lambdaRequestId = args[0].event.requestContext.requestId;
    return {
      // ...(await lambdaMiddleware(...args)),
      // transactionId,
      // logger: logger.child({
      //   service: 'api',
      //   transactionId,
      //   lambdaRequestId,
      //   referer,
      //   environment: process.env.STAGE,
      // }),
    };
  },
  debug: true,
  tracing: true,
  introspection: true,
});

export default server;
