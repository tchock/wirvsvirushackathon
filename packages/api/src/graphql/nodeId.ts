import { Kind } from 'graphql/language';
import { GraphQLScalarType } from 'graphql';

function serialize(value) {
  return Buffer.from(value).toString('base64');
}

function parse(value) {
  if (typeof value === 'string') {
    return Buffer.from(value, 'base64').toString();
  }
  throw new Error('NodeId has to be a string');
}

const NodeId = new GraphQLScalarType({
  name: 'NodeId',
  description: 'serializes and deserialized nodeIds',
  serialize,
  parseValue: parse,
  parseLiteral(ast) {
    if (ast.kind === Kind.STRING) {
      return parse(ast.value);
    }
    throw new Error('NodeId has to be a string');
  },
});

export default NodeId;
