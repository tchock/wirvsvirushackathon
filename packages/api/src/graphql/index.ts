import { makeExecutableSchema } from 'apollo-server-lambda';
import { asPaginationResolver } from './pagination';
import Order from '../Orders';
import NodeId from './nodeId';

const typeDef = /* GraphQL */ `
  type Query {
    orders(audience: Audiences!, orderStatus: OrderStatus): OrderList
    order(audience: Audiences!, nodeId: NodeId!): Order
  }

  type Mutation {
    orderPickUp(pickUpCode: String!): Order # order status -> PICKED_UP (by ??)
    orderDecline(nodeId: NodeId!, reason: String): Order # order status -> REJECTED (by store owner)
    orderAccept(nodeId: NodeId!): Order # order status -> ACCEPTED (by store owner)
    orderPlace(order: OrderInput): Order # creates order (by customer)
  }

  enum Audiences {
    CUSTOMER
    STORE
  }

  enum OrderStatus {
    PENDING
    ACCEPTED
    REJECTED
    READY_FOR_PICKUP
    PICKED_UP
  }

  interface Node {
    nodeId: NodeId! # base64 PK::SK
  }

  type OrderList {
    # edges: [OrderEdge]!
    nodes: [Order]!
  }

  # type OrderEdge {
  #   cursor: String!
  #   node: Order!
  # }

  type Order implements Node {
    nodeId: NodeId!
    pickUpCode: String!
    confirmedPickUpTime: String # ISO8601
    requestedPickUpTime: String # ISO8601
    store: Store!
    customer: Customer!
    orderStatus: OrderStatus!
    bundles: BundleList!
    shareLink: String
  }

  input OrderInput {
    bundles: [BundleInput!]!
    store: NodeId! # Store nodeId
    requestedPickUpTime: String # ISO8601
  }

  type BundleList {
    # edges: [BundleEdge]!
    nodes: [Bundle]!
  }

  # type BundleEdge {
  #   cursor: String!
  #   node: Bundle!
  # }

  type Bundle implements Node {
    nodeId: NodeId! # NOT any PK + SK
    items: BundleItemList!
  }

  input BundleInput {
    nodeId: NodeId!
    items: [BundleItemInput!]!
  }

  type BundleItemList {
    # edges: [BundleItemEdge]!
    nodes: [BundleItem]!
  }

  # type BundleItemEdge {
  #   cursor: String!
  #   node: BundleItem!
  # }

  type BundleItem {
    nodeId: NodeId! # NOT any PK + SK
    quantity: Float!
    price: Float!
    name: String!
    unit: String!
  }

  input BundleItemInput {
    nodeId: NodeId! # NOT any PK + SK
    quantity: Float!
    # once we handle bundles on the server side we should drop all these for placing an order
    price: Float!
    name: String!
    unit: String!
  }

  interface User {
    type: Audiences!
  }

  type Store implements Node & User {
    nodeId: NodeId!
    type: Audiences!
  }

  type Customer implements Node & User {
    nodeId: NodeId!
    type: Audiences!
  }

  scalar NodeId
`;

const resolvers = {
  Query: {
    orders: asPaginationResolver(Order.orders),
    order: Order.order,
  },
  Mutation: {
    orderPlace: Order.orderPlace,
    orderAccept: Order.orderAccept,
    orderDecline: Order.orderDecline,
    orderPickUp: Order.orderPickUp,
  },
  Order: {
    bundles: asPaginationResolver(parent => parent.bundles),
  },
  Bundle: {
    items: asPaginationResolver(parent => parent.items),
  },
  NodeId,
};

export const schema = makeExecutableSchema({ typeDefs: [typeDef], resolvers });
