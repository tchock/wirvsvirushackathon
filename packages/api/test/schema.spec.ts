import { printSchema } from 'graphql';
import { schema } from '../src/graphql';

test(`exported schema`, async () => {
  expect(printSchema(schema)).toMatchSnapshot();
});
