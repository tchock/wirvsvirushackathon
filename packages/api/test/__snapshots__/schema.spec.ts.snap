// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exported schema 1`] = `
"enum Audiences {
  CUSTOMER
  STORE
}

type Bundle implements Node {
  nodeId: NodeId!
  items: BundleItemList!
}

input BundleInput {
  items: BundleItemInput!
}

type BundleItem {
  nodeId: NodeId!
  quantity: Float!
  price: Float!
  name: String!
  unit: String!
}

input BundleItemInput {
  nodeId: NodeId!
  quantity: Float!
}

type BundleItemList {
  \\"\\"\\"edges: [BundleItemEdge]!\\"\\"\\"
  nodes: [BundleItem]!
}

type BundleList {
  \\"\\"\\"edges: [BundleEdge]!\\"\\"\\"
  nodes: [Bundle]!
}

type Customer implements Node & User {
  nodeId: NodeId!
  type: Audiences!
}

type Mutation {
  \\"\\"\\"
  orderPickUp(pickUpCode: String!): Order # order status -> PICKED_UP (by ??)
  \\"\\"\\"
  orderDecline(nodeId: NodeId!): Order
  orderAccept(nodeId: NodeId!): Order
  orderPlace(order: OrderInput): Order
}

interface Node {
  nodeId: NodeId!
}

\\"\\"\\"serializes and deserialized nodeIds\\"\\"\\"
scalar NodeId

type Order implements Node {
  nodeId: NodeId!
  pickUpCode: String!
  confirmedPickUpTime: String
  requestedPickUpTime: String
  store: Store!
  customer: Customer!
  orderStatus: OrderStatus!
  bundles: BundleList!
  shareLink: String!
}

input OrderInput {
  bundles: [BundleInput!]!
  store: String!
  requestedPickUpTime: String
}

type OrderList {
  \\"\\"\\"edges: [OrderEdge]!\\"\\"\\"
  nodes: [Order]!
}

enum OrderStatus {
  PENDING
  ACCEPTED
  REJECTED
  READY_FOR_PICKUP
  PICKED_UP
}

type Query {
  orders(audience: Audiences!, orderStatus: OrderStatus): OrderList
  order(audience: Audiences!, nodeId: NodeId!): Order
}

type Store implements Node & User {
  nodeId: NodeId!
  type: Audiences!
}

interface User {
  type: Audiences!
}
"
`;
